### PR TITLE
Unique group id support. 

### PR DESCRIPTION
**Pull request by blublinsky**
### What changes were proposed in this pull request?

It introduces an additional streamlet type, that reads data from all partitions

### Why are the changes needed?

This effectively implements broadcast to all instances of akka streams based streamlets. This is an important features to enable scalability for dynamically controlled streams with a single key.

### Does this PR introduce any user-facing change?

It does not impact an existing code, but introduce a new API, thus allowing to control whether a streamlet reads from all partitions or using partitioning strategy based on scaling of the streamlet

### How was this patch tested?

No test are necessary. It basically creates a unique groupID for Kafka readers

### CODE REVIEW

File: core/build.sbt [hunk 1]
```diff
@@ -124,7 +122,10 @@ lazy val akkastreamTestkit =
     .settings(
       (sourceDirectory in AvroConfig) := baseDirectory.value / "src/test/avro",
       (stringType in AvroConfig) := "String",
-      javacOptions += "-Xlint:unchecked"
+      javacOptions += "-Xlint:unchecked",
+      javaOptions += "-XX:MaxMetaspaceSize=512M",
```
Comment (RayRoestenburg) [inline core/build.sbt]: Please put these changes in a separate PR, they are not related to reading from all partitions.

File: core/cloudflow-spark-tests/src/test/scala/cloudflow/spark/SparkEgressSpec.scala [hunk 2]
```diff
@@ -81,8 +81,8 @@ class SparkEgressSpec extends SparkScalaTestSupport {
       val r2 = asCollection[String](session, "allNamesUpper")
 
       // assert
-      r1 must contain("name1")
-      r2 must contain("NAME1")
+      r1 must contain("name2")
```
Comment (RayRoestenburg) [inline core/cloudflow-spark-tests/src/test/scala/cloudflow/spark/SparkEgressSpec.scala]: This also does not look like it is related to read all partitions. Why was this change necessary? Did the test fail?

File: core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/avro/AvroInlet.scala [hunk 3]
```diff
@@ -35,3 +36,18 @@ object AvroInlet {
   def create[T <: SpecificRecordBase](name: String, clazz: Class[T]): AvroInlet[T] =
     AvroInlet[T](name)(ClassTag.apply(clazz))
 }
+
+// Additional AvroInlet reading from all partitions of the topic. This is done by ensuring unique
+// groupID for Kafka listener
+case class AvroInletAllPartitions[T <: SpecificRecordBase: ClassTag](name: String) extends CodecInlet[T] {
```
Comment (RayRoestenburg) [inline core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/avro/AvroInlet.scala]: Instead it would be more consistent to add a `withUniqueGroupId` to `CodecInlet`. We already have a 'builder' API for `CodecOutlet`, which allows for setting a partitioner with `withPartitioner`. So doing it that way would be more consistent. Also, that way, this class is not necessary anymore.

**Comment by RayRoestenburg** [comment]
Also, please also look at `SparkStreamletContext` and `FlinkStreamletContext`, if you want to be able to read as a unique group there too.

**Comment by blublinsky** [comment]
I do not think we need to unique groupID in Flink and Spark. Both allow to bring all of the messages from all partitions together. This is something that is only required for Akka streams

File: core/build.sbt (hunk 1 above)
Comment (blublinsky) [reply core/build.sbt]: Did this only to be able to run tests locally, removing

File: core/cloudflow-spark-tests/src/test/scala/cloudflow/spark/SparkEgressSpec.scala (hunk 2 above)
Comment (blublinsky) [reply core/cloudflow-spark-tests/src/test/scala/cloudflow/spark/SparkEgressSpec.scala]: Done

File: core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/avro/AvroInlet.scala (hunk 3 above)
Comment (blublinsky) [reply core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/avro/AvroInlet.scala]: Redesign based on Ray's suggestion of implementing a builder

File: core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/StreamletPort.scala [hunk 4]
```diff
@@ -61,6 +61,18 @@ trait CodecInlet[T] extends Inlet {
    * Describes the schema used to deserialize the data.
    */
   def schemaAsString: String
+
+  /**
+   * Returns if this inlet uses a unique group id.
+   */
```
Comment (agemooij) [inline core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/StreamletPort.scala]: Can we please use the same, more informative comment as used for the Java version below?

File: core/cloudflow-akka/src/main/scala/cloudflow/akkastream/AkkaStreamletContextImpl.scala [hunk 5]
```diff
@@ -73,7 +74,11 @@ final class AkkaStreamletContextImpl(
   }
 
   private val bootstrapServers = system.settings.config.getString("cloudflow.kafka.bootstrap-servers")
-  private def groupId[T](savepointPath: SavepointPath, streamletRef: String, inlet: CodecInlet[T]) = s"${savepointPath.appId}.${streamletRef}.${inlet.name}"
+  private def groupId[T](savepointPath: SavepointPath, streamletRef: String, inlet: CodecInlet[T]) = {
+    val base = s"${savepointPath.appId}.${streamletRef}.${inlet.name}"
+    if (inlet.hasUniqueGroupId) base + randomUUID.toString
+    else base
+  }
```
Comment (agemooij) [inline core/cloudflow-akka/src/main/scala/cloudflow/akkastream/AkkaStreamletContextImpl.scala]: Are there any length restrictions in group IDs we need to worry about here? A GUID is quite long...

Comment (RayRoestenburg) [reply core/cloudflow-akka/src/main/scala/cloudflow/akkastream/AkkaStreamletContextImpl.scala]: No:
https://github.com/apache/kafka/blob/72282ed1988e33000f91672a72c7356e26658241/core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala#L799

File: core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/StreamletPort.scala (hunk 4 above)
Comment (RayRoestenburg) [reply core/cloudflow-streamlets/src/main/scala/cloudflow/streamlets/StreamletPort.scala]: The one returns the boolean, the other sets it, but yes, I'll add it.
